File: app/src/main/java/com/example/aleartmycontroller/data/repository/ObservationEventRepository.kt
package com.example.aleartmycontroller.data.repository

import com.example.aleartmycontroller.data.local.dao.ObservationEventDao
import com.example.aleartmycontroller.data.local.entity.EventEntity
import com.example.aleartmycontroller.data.local.entity.ObservationEventEntity
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ObservationEventRepository @Inject constructor(
    private val dao: ObservationEventDao
) {
    /**
     * EventEntity に対応する ObservationEvent の obsEventId を返す。
     * 既存行があればその ID を返し、なければスナップショットを新規挿入して ID を返す。
     *
     * INSERT OR IGNORE の仕様上、既存行が無視された場合は -1 が返るため、
     * その場合は改めて SELECT で取得する。
     */
    suspend fun findOrCreate(event: EventEntity): Long {
        val inserted = dao.insert(
            ObservationEventEntity(
                googleEventId = event.googleEventId,
                title = event.title,
                startTime = event.startTime,
                endTime = event.endTime
            )
        )
        if (inserted != -1L) return inserted
        return dao.findByGoogleEventId(event.googleEventId)!!.obsEventId
    }

    suspend fun findById(obsEventId: Long): ObservationEventEntity? =
        dao.findById(obsEventId)
}

File: flutter/amc/lib/data/repository/observation_event_repository.dart
import 'package:drift/drift.dart';

import '../local/daos/observation_event_dao.dart';
import '../local/database.dart';

/// ObservationEvent の findOrCreate を担う（Android: ObservationEventRepository 相当）。
class ObservationEventRepository {
  ObservationEventRepository(this._dao);

  final ObservationEventDao _dao;

  /// Event(カレンダーキャッシュ) に対応する ObservationEvent の obsEventId を返す。
  ///
  /// 不変条件(§9): insert-or-ignore のセマンティクス。重複 googleEventId では
  /// 既存の obsEventId を返す。drift の insertOrIgnore は戻り値が信頼できないため、
  /// 常に [ObservationEventDao.findByGoogleEventId] で解決する。
  Future<int> findOrCreate(Event event) async {
    await _dao.insertOrIgnore(
      ObservationEventsCompanion.insert(
        googleEventId: Value(event.googleEventId),
        title: event.title,
        startTime: event.startTime,
        endTime: event.endTime,
      ),
    );
    final existing = await _dao.findByGoogleEventId(event.googleEventId);
    return existing!.obsEventId;
  }

  Future<ObservationEvent?> findById(int obsEventId) => _dao.findById(obsEventId);

  /// cloud pull 専用: 既存行があれば返し、なければ新規作成する。
  /// Event row を持たないクラウド由来レコードに使う（Event.googleEventId の代わりに生の文字列で解決）。
  Future<int> findOrCreateByRaw({
    required String googleEventId,
    required String title,
    required int startTime,
    required int endTime,
  }) async {
    final existing = await _dao.findByGoogleEventId(googleEventId);
    if (existing != null) return existing.obsEventId;
    await _dao.insertOrIgnore(
      ObservationEventsCompanion.insert(
        googleEventId: Value(googleEventId),
        title: title,
        startTime: startTime,
        endTime: endTime,
      ),
    );
    final created = await _dao.findByGoogleEventId(googleEventId);
    return created!.obsEventId;
  }
}
